File: kirun-js/__tests__/engine/function/system/array/FillTest.ts
import { Fill } from '../../../../../src/engine/function/system/array/Fill';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';
import { MapUtil } from '../../../../../src/engine/util/MapUtil';

test('Fill Test', async () => {
    let fill: Fill = new Fill();

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters();
    let array = [0, 1];

    fep.setArguments(MapUtil.of('source', array as any, 'element', 3));

    await fill.execute(fep);
    expect(array).toStrictEqual([3, 3]);

    fep.setArguments(MapUtil.of('source', array as any, 'element', 5, 'srcFrom', 2, 'length', 5))
        .setContext(MapUtil.of())
        .setSteps(MapUtil.of());

    let finArray = [3, 3, 5, 5, 5, 5, 5];
    await fill.execute(fep);

    expect(array).toStrictEqual(finArray);

    fep.setArguments(MapUtil.of('source', array as any, 'element', 25, 'srcFrom', 5));

    finArray = [3, 3, 5, 5, 5, 25, 25];

    await fill.execute(fep);

    expect(array).toStrictEqual(finArray);

    fep.setArguments(MapUtil.of('source', array as any, 'element', 20, 'srcFrom', -1));

    await expect(fill.execute(fep)).rejects.toThrow();
});

File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/array/FillTest.java
package com.fincity.nocode.kirun.engine.function.system.array;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class FillTest {

	@Test
	void testInternalExecute() {

		Fill fill = new Fill();

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());
		JsonArray array = new JsonArray();
		array.add(0);
		array.add(1);

		fep.setArguments(Map.of("source", array, "element", new JsonPrimitive(3)))
				.setContext(Map.of())
				.setSteps(Map.of());

		fill.execute(fep).block();

		JsonArray finArray = new JsonArray();
		finArray.add(new JsonPrimitive(3));
		finArray.add(new JsonPrimitive(3));

		assertEquals(finArray, array);

		fep.setArguments(Map.of("source", array, "element", new JsonPrimitive(5), "srcFrom", new JsonPrimitive(2),
				"length", new JsonPrimitive(5)))
				.setContext(Map.of())
				.setSteps(Map.of());

		finArray = new JsonArray();
		finArray.add(new JsonPrimitive(3));
		finArray.add(new JsonPrimitive(3));
		finArray.add(new JsonPrimitive(5));
		finArray.add(new JsonPrimitive(5));
		finArray.add(new JsonPrimitive(5));
		finArray.add(new JsonPrimitive(5));
		finArray.add(new JsonPrimitive(5));

		fill.execute(fep).block();

		assertEquals(finArray, array);

		fep.setArguments(Map.of("source", array, "element", new JsonPrimitive(25), "srcFrom", new JsonPrimitive(5)))
				.setContext(Map.of())
				.setSteps(Map.of());

		finArray = new JsonArray();
		finArray.add(new JsonPrimitive(3));
		finArray.add(new JsonPrimitive(3));
		finArray.add(new JsonPrimitive(5));
		finArray.add(new JsonPrimitive(5));
		finArray.add(new JsonPrimitive(5));
		finArray.add(new JsonPrimitive(25));
		finArray.add(new JsonPrimitive(25));

		fill.execute(fep).block();

		assertEquals(finArray, array);

		fep.setArguments(Map.of("source", array, "element", new JsonPrimitive(20), "srcFrom", new JsonPrimitive(-1)))
				.setContext(Map.of())
				.setSteps(Map.of());

		StepVerifier.create(fill.execute(fep))
				.expectError(KIRuntimeException.class)
				.verify();
	}

}
